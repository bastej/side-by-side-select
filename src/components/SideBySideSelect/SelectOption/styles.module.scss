.SelectOption {
  padding: 3px 12px;
  cursor: pointer;
  &:hover {
    background-color: #eee;
  }
  &.Marked {
    background-color: #eee;
  }
  &.Disabled {
    cursor: not-allowed;
    color: #aaa;
    background-color: #ddd;
  }
}
