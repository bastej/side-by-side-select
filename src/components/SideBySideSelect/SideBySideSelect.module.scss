.SideBySideSelect {
  display: flex;
}

.ButtonContainer {
  display: flex;
  padding: 8px;
  flex-direction: column;
  justify-content: center;
}

.Button {
  margin: 3px 0;
  border: 1px solid #ccc;
}

.List {
  max-height: 200px;
  padding: 0;
  list-style: none;
  overflow-y: auto;
  margin-bottom: 0;
}
