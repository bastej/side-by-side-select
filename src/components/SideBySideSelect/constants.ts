export const ALL_OPTIONS_SIDE = 'ALL_OPTIONS_SIDE';
export const CHOSEN_OPTIONS_SIDE = 'CHOSEN_OPTIONS_SIDE';

export const selectedItemsDefaultHeader = 'Selected';
export const allItemsDefaultHeader = 'All';
