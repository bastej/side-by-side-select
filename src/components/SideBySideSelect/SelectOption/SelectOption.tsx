import React, { MouseEventHandler, ReactNode } from 'react';
import { disabledItemDefaultTitle } from './constants';

import styles from './styles.module.scss';

export type Props = {
  onClick: MouseEventHandler<HTMLLIElement>;
  onDoubleClick: MouseEventHandler<HTMLLIElement>;
  isMarked?: boolean;
  isDisabled?: boolean;
  disabledItemTitle?: string;
  children: ReactNode;
};

const SelectOption = ({
  onClick,
  onDoubleClick,
  isMarked = false,
  isDisabled = false,
  disabledItemTitle = disabledItemDefaultTitle,
  children,
}: Props): JSX.Element => {
  const dependentProps = {
    onClick,
    onDoubleClick,
  };

  return (
    <li
      className={`${styles.SelectOption} ${isMarked ? styles.Marked : ''} ${isDisabled ? styles.Disabled : ''}`}
      {...(isDisabled ? { title: disabledItemTitle } : { ...dependentProps })}
    >
      {children}
    </li>
  );
};

export default SelectOption;
