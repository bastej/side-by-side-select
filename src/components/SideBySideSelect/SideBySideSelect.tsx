import React, { useState, useEffect, HTMLAttributes } from "react";

import styles from "./SideBySideSelect.module.scss";

import SelectOption from "./SelectOption/SelectOption";
import SelectSide from "./SelectSide/SelectSide";
import SelectActionButton from "./SelectActionButton/SelectActionButton";
import {
  allItemsDefaultHeader,
  ALL_OPTIONS_SIDE,
  CHOSEN_OPTIONS_SIDE,
  selectedItemsDefaultHeader,
} from "./constants";

type ActiveSide = typeof ALL_OPTIONS_SIDE | typeof CHOSEN_OPTIONS_SIDE | "";
type SideBySideSelectTexts = {
  selectedItemsHeader: string;
  allItemsHeader: string;
};

export type Props<Type> = {
  label: string;
  allOptions: Type[];
  disabledOptions?: Type[];
  initialSelectedOptions?: Type[];
  errorOccurred?: boolean;
  rootElementProps?: HTMLAttributes<HTMLDivElement>;
  sortOptions?: boolean;
  texts: SideBySideSelectTexts;
  onChange(value: Type[]): void;
};

function SideBySideSelect<T>({
  label,
  allOptions,
  disabledOptions = [],
  initialSelectedOptions = [],
  errorOccurred = false,
  onChange,
  rootElementProps,
  sortOptions = false,
  texts: {
    selectedItemsHeader = selectedItemsDefaultHeader,
    allItemsHeader = allItemsDefaultHeader,
  },
}: Props<T>) {
  const [selectedItems, setSelectedItems] = useState<T[]>(initialSelectedOptions);
  const [markedItems, setMarkedItems] = useState<T[]>([]);
  const [activeSide, setActiveSide] = useState<ActiveSide>("");

  useEffect(() => {
    onChange(selectedItems);
  }, [selectedItems]);

  useEffect(() => {
    setMarkedItems([]);
  }, [activeSide]);

  function selectItems(items: T[]): void {
    const updatedSelectedItems = [...selectedItems, ...items];

    setSelectedItems(updatedSelectedItems);
    setMarkedItems([]);
  }

  function deselectItems(items: T[]) {
    const updatedSelectedItems =
      selectedItems && selectedItems.filter((el: T) => !items.includes(el));

    setSelectedItems(updatedSelectedItems);
    setMarkedItems([]);
  }

  function markItem(item: T): void {
    setMarkedItems([...markedItems, item]);
  }

  function unmarkItem(item: T): void {
    const updatedItems: T[] = markedItems.filter((el: T) => el !== item);

    setMarkedItems(updatedItems);
  }

  function renderAllItem(): JSX.Element {
    const itemsToShow = allOptions && allOptions.filter((item) => !selectedItems.includes(item));

    if (sortOptions && itemsToShow) {
      itemsToShow.sort();
    }

    return (
      <ul className={styles.List}>
        {itemsToShow &&
          itemsToShow.map((item: T, i) => {
            const isMarked: boolean = markedItems.includes(item);
            const isDisabled: boolean = disabledOptions.includes(item);

            return (
              <SelectOption
                key={`${item}_${i}`}
                onClick={(): void => {
                  if (isMarked) {
                    unmarkItem(item);
                  } else {
                    markItem(item);
                  }
                }}
                onDoubleClick={() => selectItems([item])}
                isMarked={isMarked}
                isDisabled={isDisabled}
              >
                {item}
              </SelectOption>
            );
          })}
      </ul>
    );
  }

  function renderSelectedItems(): JSX.Element {
    const itemsToShow = selectedItems;

    if (sortOptions && itemsToShow) {
      itemsToShow.sort();
    }

    return (
      <ul className={styles.List}>
        {selectedItems &&
          selectedItems.map((item: T, i) => {
            const isMarked = markedItems.includes(item);

            return (
              <SelectOption
                key={`${item}_${i}`}
                onClick={(): void => {
                  if (isMarked) {
                    unmarkItem(item);
                  } else {
                    markItem(item);
                  }
                }}
                onDoubleClick={() => deselectItems([item])}
                isMarked={isMarked}
              >
                {item}
              </SelectOption>
            );
          })}
      </ul>
    );
  }

  return (
    <div {...rootElementProps}>
      {label}
      <div className={styles.SideBySideSelect}>
        <SelectSide
          onClick={(): void => {
            setActiveSide(ALL_OPTIONS_SIDE);
          }}
          isActive={activeSide === ALL_OPTIONS_SIDE}
          header={allItemsHeader}
          errorOccurred={errorOccurred}
          testID="left-side"
        >
          {renderAllItem()}
        </SelectSide>
        <div className={styles.ButtonContainer}>
          <SelectActionButton
            onClick={(): void => {
              selectItems(markedItems);
            }}
            content=">"
            disabled={!!(markedItems.length === 0 || activeSide !== ALL_OPTIONS_SIDE)}
            testID="moveToRightButton"
          />
          <SelectActionButton
            onClick={(): void => {
              // select all items except disabled and already added ones
              const itemsToShow: T[] =
                allOptions &&
                allOptions.filter(
                  (item) => !disabledOptions.includes(item) && !selectedItems.includes(item)
                );

              selectItems(itemsToShow);
            }}
            content=">>"
            testID="moveAllToRightButton"
          />
          <SelectActionButton
            onClick={(): void => {
              deselectItems(markedItems);
            }}
            content="<"
            disabled={!!(markedItems.length === 0 || activeSide !== CHOSEN_OPTIONS_SIDE)}
            testID="moveToLeftButton"
          />
          <SelectActionButton
            onClick={(): void => {
              setSelectedItems([]);
            }}
            content="<<"
            testID="moveAllToLeftButton"
          />
        </div>
        <SelectSide
          onClick={(): void => {
            setActiveSide(CHOSEN_OPTIONS_SIDE);
          }}
          isActive={activeSide === CHOSEN_OPTIONS_SIDE}
          header={selectedItemsHeader}
          errorOccurred={errorOccurred}
          testID="right-side"
        >
          {renderSelectedItems()}
        </SelectSide>
      </div>
    </div>
  );
}

export default SideBySideSelect;
