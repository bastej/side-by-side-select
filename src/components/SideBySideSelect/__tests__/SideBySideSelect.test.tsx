/**
 * '>' - select marked item (move to selected item list)
 * '>>' - select all items (move to selected item list)
 * '<' - unselect marked item (move to items to select list)
 * '<<' - unselect all items (move to items to select item list)
 */

import React from "react";
import { create } from "react-test-renderer";
import { render } from "@testing-library/react";
import { disabledItemDefaultTitle } from "../SelectOption/constants";

import SideBySideSelect, { Props } from "../SideBySideSelect";

jest.mock("../../SelectOption/styles.module.scss", () => ({
  Disabled: "disabled",
}));

describe("<SideBySideSelect />", () => {
  const commonProps: Props = {
    texts: {
      allItemsHeader: "all",
      selectedItemsHeader: "selected",
    },
    label: "Field label",
    onChange: jest.fn(),
    allOptions: ["option1", "option2", "option3", "option4"],
  };

  it("renders and match snapshot", () => {
    const tree = create(<SideBySideSelect {...commonProps} />);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("renders options with specific class and title for disabled", () => {
    const disabledOptions = ["option1"];
    const root = render(<SideBySideSelect {...commonProps} disabledOptions={disabledOptions} />);

    const option = root.getByText("option1");

    expect(option.classList.contains("disabled")).toBe(true);
    expect(option.getAttribute("title")).toBe(disabledItemDefaultTitle);
  });

  it("renders select with initially selected item", () => {
    const root = render(<SideBySideSelect {...commonProps} initialSelectedOptions={["option1"]} />);

    const leftSide = root.getByTestId("left-side");
    const rightSide = root.getByTestId("right-side");

    expect(leftSide.contains(root.getByText("option1"))).toBe(false);
    expect(rightSide.contains(root.getByText("option1"))).toBe(true);
  });
});
