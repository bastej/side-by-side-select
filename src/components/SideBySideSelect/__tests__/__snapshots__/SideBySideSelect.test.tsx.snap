// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideBySideSelect /> renders and match snapshot 1`] = `
<div>
  Field label
  <div>
    <div
      className="undefined "
      data-testid="left-side"
      onClick={[Function]}
    >
      <div
        data-testid="side-header"
        style={Object {}}
      >
        all
      </div>
      <ul>
        <li
          className="undefined  "
          onClick={[Function]}
          onDoubleClick={[Function]}
        >
          option1
        </li>
        <li
          className="undefined  "
          onClick={[Function]}
          onDoubleClick={[Function]}
        >
          option2
        </li>
        <li
          className="undefined  "
          onClick={[Function]}
          onDoubleClick={[Function]}
        >
          option3
        </li>
        <li
          className="undefined  "
          onClick={[Function]}
          onDoubleClick={[Function]}
        >
          option4
        </li>
      </ul>
    </div>
    <div>
      <button
        className="undefined disabled"
        data-testid="moveToRightButton"
        disabled={true}
        onClick={[Function]}
      >
        &gt;
      </button>
      <button
        className="undefined "
        data-testid="moveAllToRightButton"
        disabled={false}
        onClick={[Function]}
      >
        &gt;&gt;
      </button>
      <button
        className="undefined disabled"
        data-testid="moveToLeftButton"
        disabled={true}
        onClick={[Function]}
      >
        &lt;
      </button>
      <button
        className="undefined "
        data-testid="moveAllToLeftButton"
        disabled={false}
        onClick={[Function]}
      >
        &lt;&lt;
      </button>
    </div>
    <div
      className="undefined "
      data-testid="right-side"
      onClick={[Function]}
    >
      <div
        data-testid="side-header"
        style={Object {}}
      >
        selected
      </div>
      <ul />
    </div>
  </div>
</div>
`;
